### PR TITLE
fix issues with overlapping areas rendering incorrectly

### PR DESCRIPTION
**Pull request by teykamp**
fixes #46

### CODE REVIEW

File: src/sets/draw/overlaps.ts [hunk 1]
```diff
@@ -49,14 +49,46 @@ type ColorOverlappingAreasProps = {
   highlightedOverlaps: Map<Overlap['id'], string[]>,
 }
 
+const getProperNonEmptySubsets = (labels: Circle['label'][]): Circle['label'][][] => {
+  const subsets: Circle['label'][][] = []
+  for (let mask = 1; mask < (2 ** labels.length) - 1; mask++) {
+    subsets.push(labels.filter((_, i) => mask & (1 << i)))
```
Comment (Yonava) [inline src/sets/draw/overlaps.ts:55]: Something like this is probably best split up into a few lines, with descriptive naming and a comment to boot

Comment (Yonava) [reply src/sets/draw/overlaps.ts:55]: Recommended:

```ts

const getProperNonEmptySubsets = (labels: Circle['label'][]): Circle['label'][][] => {
  const subsets: Circle['label'][][] = []
  const fullSetMask = 2 ** labels.length - 1

  // Each mask from 1 to fullSetMask - 1 represents one proper, non-empty subset,
  // where bit i indicates whether labels[i] is included.
  for (let mask = 1; mask < fullSetMask; mask++) {
    const subset = labels.filter((_, i) => mask & (1 << i))
    subsets.push(subset)
  }

  return subsets
}
```

File: src/sets/draw/overlaps.ts [hunk 2]
```diff
@@ -49,14 +49,46 @@ type ColorOverlappingAreasProps = {
...
+  for (const overlap of overlaps) {
+    const key = overlap.circles.toSorted((a, b) => a.localeCompare(b)).join('.')
+    overlapIdByKey.set(key, overlap.id)
+  }
+
+  // an overlap region is geometrically nested inside every region formed by a
+  // subset of its circles, so it must be redrawn (even just to erase it back to
+  // background) whenever a single circle or a smaller
+  // overlap is highlighted, otherwise that region's fill bleeds through
+  const hasHighlightedAncestor = (labels: Circle['label'][]) => {
```
Comment (Yonava) [inline src/sets/draw/overlaps.ts:76]: All this code added is perfect for unit tests. Pure functions. Easy correct/incorrect. Dense enough that edge cases and regressions can easily sneak by.

File: src/sets/draw/overlaps.ts [hunk 3]
```diff
@@ -49,14 +49,46 @@ type ColorOverlappingAreasProps = {
...
+    for (const subset of getProperNonEmptySubsets(labels)) {
+      if (subset.length === 1) {
+        if (highlightedCircles.has(subset[0])) return true
+        continue
+      }
+      const key = subset.toSorted((a, b) => a.localeCompare(b)).join('.')
```
Comment (Yonava) [inline src/sets/draw/overlaps.ts:82]:
```suggestion
      const key = subset.toSorted().join('.')
```

Comment (Yonava) [inline src/sets/draw/overlaps.ts:82]: Key is also a toStringed subset. So instead of having all this logic inline. I would just define a helper at the top called subsetToString, do the logic there, and dont name this variable `key`, name it what it is. The string subset

File: src/sets/draw/overlaps.ts (hunk 2 above)
Comment (teykamp) [reply src/sets/draw/overlaps.ts:76]: tests to come 👍 🧪
